src/charm.py

```diff
@@ -121,92 +126,116 @@ def _get_database_relation_info(self) -> dict:
             "endpoints": relation_data["endpoints"],
         }
 
-    def _update_layer(self) -> None:
-        """Updates the Pebble configuration layer and config if changed."""
-        self.unit.status = MaintenanceStatus("Applying pebble layer")
+    def _run_sql_migration(self, set_timeout) -> None:
+        """Runs a command to create SQL schemas and apply migration plans."""
+        process = self._container.exec(
+            ["hydra", "migrate", "sql", "-e", "--config", self._hydra_config_path, "--yes"],
+            timeout=20.0 if set_timeout else None,
+        )
+        try:
+            stdout, _ = process.wait_output()
+            logger.info(f"Executing automigration: {stdout}")
+        except ExecError as err:
+            logger.error(f"Exited with code {err.exit_code}. Stderr: {err.stderr}")
+            self.unit.status = BlockedStatus("Database migration job failed")
+            logger.error("Automigration job failed, please use the run-migration action")
 
-        if not self._container.get_plan().to_dict():
-            # Push the config on first layer update to avoid PathError
-            self._container.push(self._hydra_config_path, self._config, make_dirs=True)
-            logger.info("Pushed hydra config")
+    def _on_hydra_pebble_ready(self, event) -> None:
+        """Event Handler for pebble ready event."""
+        if not self.model.relations["pg-database"]:
+            logger.error("Missing required relation with postgresql")
+            self.model.unit.status = BlockedStatus("Missing required relation with postgresql")
+            return
+
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Hydra container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
+            return
+
+        self.unit.status = MaintenanceStatus("Configuring resources")
 
         self._container.add_layer(self._container_name, self._hydra_layer, combine=True)
         logger.info("Pebble plan updated with new configuration, replanning")
-
         try:
             self._container.replan()
         except ChangeError as err:
             logger.error(str(err))
             self.unit.status = BlockedStatus("Failed to replan")
             return
+        finally:
+            if self.database.is_database_created():
+                if not self._container.exists(self._hydra_config_path):
+                    logger.info(
+                        "Config file not found. Pushing it to the container and replanning"
+                    )
+                    self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+                    self._container.replan()
+                self.unit.status = ActiveStatus()
```

Comment:
This could be rewritten as:
```suggestion
        self._container.push(self._hydra_config_path, self._config, make_dirs=True)
        try:
            self._container.replan()
        except ChangeError as err:
            logger.error(str(err))
            self.unit.status = BlockedStatus("Failed to replan")
            return
        self.unit.status = ActiveStatus()
```
This way we are pushing the config file every time (which is not needed), but the flow is much simpler. What do you think?